worker/codalabworker/aws_batch.py

```diff
@@ -0,0 +1,539 @@
+                        'value': self.docker_working_directory
+                    }
+                ],  # TODO Should the env be set?
+                'mountPoints': map(lambda pair: pair[1], volumes_and_mounts),
+                'readonlyRootFilesystem': False,
+                'user': 'root'  # TODO Figure out what to do here
+            },
+            'retryStrategy': {
+                'attempts': 1
+            }
+        }
+
+        self.logger.debug(job_definition)
+
+        response = self._batch_client.register_job_definition(**job_definition)
+        arn = response['jobDefinitionArn']
+
+        self.logger.debug("Job %s registered job definition arn %s", self.uuid, arn)
+        return arn
+
+    def volume_and_mount(self, host_path, container_path, name, read_only):
+        volume_definition = {
+            'host': {
+                # TODO Figure out how to break assumption about shared file system
+                'sourcePath': host_path
+            },
+            'name': name
+        }
+
+        mount_point = {
+            'containerPath': container_path,
+            'readOnly': read_only,
+            'sourceVolume': name
+        }
+
+        return volume_definition, mount_point
+
+
+class Submit(AwsBatchRunState):
+    def update(self):
+        job_definition = self._bundle['metadata']['batch_job_definition']
+
+        # Submit job to AWS Batch
+        response = self._batch_client.submit_job(
+            jobName=self.uuid,
+            jobQueue=self.batch_queue,
+            jobDefinition=job_definition
+        )
+
+        job_id = response['jobId']
+
+        self.update_metadata(batch_job_id=job_id, submit_time=current_time())
+
+        return self.transition(Running)
+
+
+class JobFetcher(object):
+    def __init__(self, job_id, batch_client):
+        self._last_check_time = 0
+        self._check_frequency = 30  # seconds
+        self._job = None
+        self._job_id = job_id
+        self._batch_client = batch_client
+
+    def should_refresh(self):
+        now = current_time()
+        if now - self._last_check_time > self._check_frequency:
+            self._last_check_time = now
+            return True
+        return False
+
+    def fetch_job(self):
+        if self.should_refresh() or self._job is None:
+            response = self._batch_client.describe_jobs(jobs=[self._job_id])
+            self._job = response['jobs'][0]
+        return self._job
+
+
+class Running(AwsBatchRunState):
+    def __init__(self, *args, **kwargs):
+        super(Running, self).__init__(*args, **kwargs)
+        self._job_fetcher = JobFetcher(self._bundle['metadata']['batch_job_id'], self._batch_client)
+
+    """
+    Waiting for Batch to schedule the job
+    """
+    def update(self):
+        job = self._job_fetcher.fetch_job()
+        status = job['status']
+
+        now = current_time()
+        started = job.get('startedAt')
+        runtime = now - (started / 1000) if started else 0
```

Comment:
Might be nice to use datetime & timedelta here, too, and just convert when passing it to update_metadata

Reply:
I decided not to do this, because the conversion to and from everywhere seemed like it would be even more code. Really this change should be made to allow datetime to be used rather than epoch time, and until then I think it will be more trouble than it is worth to use the proper types here.